### DISCUSSION TITLE
Construct a Doc object from dataset with sent_starts  and deps/heads

### DISCUSSION THREAD

**Discussion by d5555** [explosion/spaCy#7420] [Help: Coding & Implementations]
I'm trying to constructing  a `Doc` object from dataset with  `words`, `spaces`, `sent_starts`, `deps` and `heads`. 
`doc = Doc(nlp.vocab, words=words, sent_starts= sent_starts, spaces=spaces, heads=heads, deps = deps)`
After assigning heads/deps,  Doc object  changes sent_starts and I cannot reassign sent_starts , otherwise I get error/warning that Doc was already parsed. 
How can I reassign sent_starts after Doc was parsed? 

I think we should give priority to `sent_starts` if it is provided when creating Doc object or set `sent_starts`  to `None` when we want to prioritize parser

## Your Environment
- Operating System: windows
- Python Version Used: 3.6
- spaCy Version Used: 3.0.3

**Comment by adrianeboyd** [comment]
The short answer is that you can't with `Doc.__init__` and in general. you probably shouldn't, or you should be prepared for some of the doc properties and attributes to break, since they assume that sentence boundaries align with parse boundaries. I think if you try hard you might be able to get them out of sync, but we try fairly hard to prevent you from doing this.

What's the use case where you have sent starts that don't align with parse boundaries?

Or are you concerned that sentence boundaries are missing if you just have heads and deps? When you set heads and deps, sent starts are automatically assigned based on the parse trees.

**Comment by d5555** [comment]
I just need to recreate doc object according to my dataset to further work with the doc.
Maybe to train NER or CATS when I already have tags and deps included into Doc.  
 I can't do it because parser changes sentence boundaries and the doc doesn't correspond  dataset anymore.  For instance the parser would split whitespace token  into a separate sentence and I get Doc object with more sentences than in my dataset. 
I would better get an error if parser discovers that some heads are out of bounderies, this way i would know that smth wrong with my dataset heads.

**Comment by adrianeboyd** [comment]
Hmm, can you give a concrete example of what your data looks like and how you're loading it? I'm a bit confused from your description of having heads+deps from the dataset vs. the parser adding sentence boundaries?

**Comment by d5555** [comment]
Sorry for confusing you.   I solved the problem in a different way. 
I guess there shouldn't be any discrepancy between heads and sent_starts in spaCy doc. 
I was forcing the parser to assign unrelated  depencency tags in one sentence from my not finished dataset (only for visualizing purporses to edit later).
It was split into 2 sentences where I wanted one sentence and some NER tags were lost because spans got torn apart .

**Comment by d5555** [comment]
Hello , 
I still think it would be good to have an option to give priority to sent_starts over heads, or spacy doc should cause en error when there are heads in one sentence that are not conected to root or no root at all when we create a doc. (maybe wrong data set , but spacy doc will be created eitherway and then will cause unalignment ) 
For instance :  (example of wrong heads, see second line )
'heads': [1, 1, 1],
'heads': [**1, 0**, 1, 1],
'heads': [1, 1, 1],
In the second sentence there is not root, but it will be parsed by spacy without causing error.  Or it can be 2 roots in one sentence by mistake.   [1, 1, 3, 3]. 

Lets concatenate to make a doc ( or we can use merge docs) : 
'heads': [1, 1, 1, **4, 3,** 3, 3, 8, 8, 8],
Now after parsing we get doc that has only 2 sentences instead of 3 ( or 4 sentences in case of 2 roots).  This way we have misalignment with the other dataset info. I can not iterate through doc.sents and dataset anymore as they are mismatched.

Is there any work around ? I want to assign sent_starts after doc is parsed.
